Sprint 15

superaydumov opened the pull request:
Pull request according to Yandex Praktikum course (sprint_15).

superaydumov committed:
Added collectionView in EventCreatorViewController and set up constraints for it.
Created EventCreatorCollectionViewCell and EventCreatorSupplementaryView classes.

superaydumov committed:
Added emojis and colours to EventCreatorViewController:
- created EventCreatorCollectionViewCell class and implemented it;
- created EventCreatorSupplementaryView class and implemented it;

Refactored TrackerCreatorViewController's createTracker method (after tracker creation user jumps to TrackersViewController).

superaydumov committed:
Refactored EventCreatorViewController class to fix issue with top label (it has not been visible).

superaydumov committed:
Refactored TrackerCreatorViewController (added buttons in stackView for buttons to be at the same spot on different devices).

superaydumov committed:
Added selectedEmoji and selectedColor properties to store picked emoji and color from collectionView.

superaydumov committed:
Refactored ScheduleViewController and EventCreatorViewController classes to show sorted days in schedule's button subtitle.

superaydumov committed:
Little changes in code (ScheduleViewController).

superaydumov committed:
Added showsVerticalScrollIndicator property to collectionView in ScheduleViewController and TrackersViewController to hide vertical scroll indicator.

superaydumov committed:
Refactored EventCreatorViewController to scroll all screen instead of scrolling collectionView only.

superaydumov committed:
Added datePicker dimensions in TrackersViewController for datePicker to look like template in Figma.
Added code in EventCreatorViewController to save UISwitch state when revisiting ScheduleViewController.
Little improvements in ScheduleViewController (created updatePerformButton method).

superaydumov committed:
Added color and emoji selection conditions to create tracker in EventCreatorViewController class.

superaydumov committed:
Refactored WeekDay enumeration to sort days by numberOfDays property and implemented sorting in ScheduleViewController.

superaydumov committed:
Created Tracker data model to store database model.
Added DatabaseManager class, moved persistentContainer property and saveContext method in this class.
refactored AppDelegate and SceneDelegate classes.

superaydumov committed:
Created TrackerStore, TrackerCategoryStore and TrackerRecordStore classes to work with CoreData.

superaydumov committed:
Updated configureCell method in TrackersCollectionViewCell class to show completed button with alpha component according to Figma template

superaydumov committed:
Added extensions for String and UIColor to convert them into each other while using CoreData.

superaydumov committed:
Changed color property type  in TrackerCoreData.

superaydumov committed:
Changed schedule property type in TrackerCoreData.

superaydumov committed:
Revert "Changed schedule property type in TrackerCoreData."

This reverts commit f238fb16f968b47fb669f96f173f244944e1f9ad.

superaydumov committed:
Added custom class [String] to TrackerCoreData's schedule property.

superaydumov committed:
Added DatabaseError enumeration to handle database errors.
Refactored TrackerStore and TrackerRecordStore classes (added methods to work with CoreData, fetch and save data).

superaydumov committed:
Refactored TrackerCategoryStore (added properties and methods to work with CoreData).
Added TrackerCategoryStoreDelegate to connect TrackerCategoryStore with TrackersViewController.
Added case to DatabaseError enumeration to handle error while fetching from CoreData.
Added TrackerCategoryStoreUpdate structure.

superaydumov committed:
Added predicateFetch method to TrackerCategoryStore to handle downloading trackers from database when searchText is not empty.
Updated trackerCategory model (added visibleTrackers method) to search through visible trackers and show them.
Refactored TrackersViewController to work with coreData.

protos17 commented on Tracker/Trackers/EventCreatorViewController/EventCreatorCollectionView/EventCreatorCollectionViewCell.swift:
```diff
@@ -0,0 +1,54 @@
+    
+    // MARK: - Computed proprties
+    
+    lazy var emojiLabel: UILabel = {
+        let label = UILabel()
+        label.font = .systemFont(ofSize: 32)
+        label.translatesAutoresizingMaskIntoConstraints = false
+        
+        return label
+    }()
+    
+    lazy var colorView: UIView = {
```
Рекомендую сделать их приватными, а при необходимости обращения к ним извне, делать это через методы класса

protos17 commented on Tracker/Trackers/EventCreatorViewController/Schedule/ScheduleViewController.swift:
```diff
@@ -86,6 +88,16 @@ final class ScheduleViewController: UIViewController {
+    
     // MARK: - Obj-C methods
     
     @objc func performButtonDidTap(sender: AnyObject) {
```
Рекомендую указать как private

superaydumov replied:
Done

protos17 commented on Tracker/Trackers/EventCreatorViewController/Schedule/ScheduleViewController.swift:
```diff
@@ -103,10 +115,14 @@ extension ScheduleViewController: UITableViewDataSource {
+        
         cell.delegate = self
         
         if indexPath.row == 6 {
```
Вместо сравнения с 6 правильнее будет указать:

        WeekDay.allCases.count - 1

superaydumov replied:
Done

protos17 commented on Tracker/Trackers/TrackersViewController.swift:
```diff
@@ -221,7 +232,11 @@ final class TrackersViewController: UIViewController {
     
     @objc func textFieldDidChanged(sender: AnyObject) {
```
Рекомендую указать как private

superaydumov replied:
Done

superaydumov committed:
Added extension for UIViewController to hide keyboard when user taps outside typing area.
Added textFieldShouldReturn method in TrackersViewController and EventCreatorViewController to hide keyboard when "return" button tapped.
Refactored WeekDays enumeration to fix bug with trackers creation (days were shifted).
Refactored ScheduleViewController: added scrollView to fix bug when sunday weren't available on a little screen device, also changed the way schedule array sorts.
Refactored TrackersViewController, StatisticsViewController and CategoryViewController to show plugs in the middle of the screen on a different screen sizes.
Refactored updateCategories method to be able to search through trackers no matter uppercase/lowercase letters in their names.
Also added some fixes after review.
